fastavro/json_writer.py

```diff
@@ -0,0 +1,131 @@
+import json
+
+from .schema import extract_record_type
+from .six import iteritems, btou
+from .write import acquaint_schema
+from ._write_py import validate
+
+
+def _write_json(datum, schema):
+    record_type = extract_record_type(schema)
+
+    passthrough_schemas = (
+        'null',
+        'boolean',
+        'string',
+        'int',
+        'long',
+        'float',
+        'double',
+        'enum',
+    )
+
+    if record_type in passthrough_schemas:
+        return datum
+
+    elif record_type == 'bytes':
+        return btou(datum, encoding='iso-8859-1')
+
+    elif record_type == 'fixed':
+        return btou(datum, encoding='iso-8859-1')
+
+    elif record_type == 'union':
+        best_match_index = -1
+        most_fields = -1
+        for index, candidate in enumerate(schema):
+            if validate(datum, candidate):
+                if extract_record_type(candidate) == 'record':
+                    fields = len(candidate['fields'])
+                    if fields > most_fields:
+                        best_match_index = index
+                        most_fields = fields
+                else:
+                    best_match_index = index
+                    break
+        if best_match_index < 0:
+            pytype = type(datum)
+            msg = '%r (type %s) do not match %s' % (datum, pytype, schema)
+            raise ValueError(msg)
+
+        best_match_schema = schema[best_match_index]
```

Comment:
This is exactly the part we want to avoid. In my code, I created a code generator that creates python beans with information about the types they contain. I understand for normal python dictionaries, there is no easy way of detecting which avro type a python type corresponds to, except in really obvious cases. 

This implementation is already worth merging, as it adds a lot of value fastavro currently doesn't support, but I think you could extend the specification here and define how the datum should look like in your implementation to correspond to a certain json, so the result would be deterministic. Of course, this would have to be added to documentations. 

For instance, python only has float native type, avro has double and fixed. Even worse, avro spec supports that an union can have more than 1 fixed type in the union, as long as the precision is different. 

So if fastavro provided an `AvroFixed` class to hold avro fixed items, this problem would be solved. If anyone wants to send a dict to this method, the datum should contains an `AvroFixed` instance to correspond to avro fixed type. Is this a good solution? Not sure. Would `decimal` type be enough instead?

Same for string and bytes blob. If you google it, you will see several users complaining about the official python library because how to represent avro types in python is non-deterministic. We can do better.